src/backend/distributed/planner/recursive_planning.c

```diff
@@ -455,6 +466,97 @@ ShouldRecursivelyPlanSubquery(Query *subquery, RecursivePlanningContext *context
 }
 
 
+/*
+ * ShouldRecursivelyPlanSetOperation determines whether the leaf queries of a
+ * set operations tree need to be recursively planned in order to support the
+ * (sub)query as a whole.
+ */
+static bool
+ShouldRecursivelyPlanSetOperation(Query *query, RecursivePlanningContext *context)
+{
+	PlannerRestrictionContext *filteredRestrictionContext = NULL;
+
+	SetOperationStmt *setOperations = (SetOperationStmt *) query->setOperations;
+	if (setOperations == NULL)
+	{
+		return false;
+	}
+
+	if (context->level == 0)
+	{
+		/*
+		 * We cannot push down top-level set operation. Recursively plan the
+		 * leaf nodes such that it becomes a router query.
+		 */
+		return true;
+	}
+
+	if (setOperations->op != SETOP_UNION)
```

Comment:
`DeferErrorIfUnsupportedUnionQuery ()` already does this. But, you wanted to make it more explicit?

Reply:
Yes, the name DeferErrorIfUnsupportedUnionQuery does not suggest it gives a deferred error for intersect or except